pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,438 @@
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForWorkloadNamespace(ctx, namespace.Name)
+	}
+}
+
+// namespaceUpdated is called when namespace is updated
+func namespaceUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	oldNamespace, ok := oldObj.(*v1.Namespace)
+	if oldNamespace == nil || !ok {
+		log.Warnf("namespaceUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+	newNamespace, ok := newObj.(*v1.Namespace)
+	if newNamespace == nil || !ok {
+		log.Warnf("namespaceUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+
+	_, labelPresentInOldNamespace := oldNamespace.Labels[WorkLoadNamespaceLabelKey]
+	_, labelPresentInNewNamespace := newNamespace.Labels[WorkLoadNamespaceLabelKey]
+
+	if !labelPresentInOldNamespace && labelPresentInNewNamespace {
+		log.Infof("observed new workload namespace: %v", newNamespace.Name)
+		pendingCRUpdatesObj.enqueueFeatureStateUpdatesForWorkloadNamespace(ctx, newNamespace.Name)
+	}
+}
+
+// namespaceDeleted is called when namespace is deleted
+func namespaceDeleted(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	namespace, ok := obj.(*v1.Namespace)
+	if namespace == nil || !ok {
+		log.Warnf("namespaceDeleted: unrecognized object %+v", obj)
+		return
+	}
+	log.Infof("namespace: %q is deleted", namespace.Namespace)
+
+	pendingCRUpdatesObj.lock.Lock()
+	defer pendingCRUpdatesObj.lock.Unlock()
+	delete(pendingCRUpdatesObj.namespaceUpdateMap, namespace.Name)
+}
+
+// getFeatureStates returns latest feature states from supervisor config-map
+// if failed to retrieve feature states, func returns error with empty array of FeatureState
+func getFeatureStates(ctx context.Context) ([]featurestatesv1alpha1.FeatureState, error) {
+	log := logger.GetLogger(ctx)
+	//  Retrieve SV FeatureStates configmap
+	fssConfigMap, err := k8sClient.CoreV1().ConfigMaps(supervisorFeatureStateConfigMapNamespace).Get(ctx, supervisorFeatureStatConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("failed to retrieve SV feature switch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
+		return nil, err
+	}
+
+	log.Infof("sucessfully retrieved SV feature switch state from namespace:%q with name: %q", supervisorFeatureStateConfigMapNamespace, supervisorFeatureStatConfigMapName)
+	var featureStates []featurestatesv1alpha1.FeatureState
+	for feature, state := range fssConfigMap.Data {
+		var featureState featurestatesv1alpha1.FeatureState
+		featureState.Name = feature
+		featureState.Enabled, err = strconv.ParseBool(state)
+		if err != nil {
+			log.Errorf("failed to parse feature state value: %v for feature: %q. Defaulting to false", state, feature)
```

Comment:
Looks like we ignore error here and move to next steps. Potentially will this cause any NPE ?
If ParseBool fails, don't we have to set `featureState.Enabled` = `false`.

Reply:
If ParseBool fails it retuns false with error, so we are good.

```
// ParseBool returns the boolean value represented by the string.
// It accepts 1, t, T, TRUE, true, True, 0, f, F, FALSE, false, False.
// Any other value returns an error.
func ParseBool(str string) (bool, error) {
	switch str {
	case "1", "t", "T", "true", "TRUE", "True":
		return true, nil
	case "0", "f", "F", "false", "FALSE", "False":
		return false, nil
	}
	return false, syntaxError("ParseBool", str)
}
```

Reply:
I would suggest to crash CSI in such cases to surface such problems sooner. Ignoring errors is problematic as system continues to work with stale data and logs getting rolled over.